### ISSUE TITLE
Odd Event Behaviour when closing a Window

### ISSUE THREAD

**Issue by shrewdlogarithm** [atom/atom-shell#1223]
Tested with the latest release of atom-shell on Win32

If I setup a 'close' event, the Window doesn't close (even if I "return true") - e.g.

````
var win = remote.getCurrentWindow()
win.close = function() {
    console.log("closing")
    // cept we never do
}
````

Also - if I close the Window using the 'right click on the title bar' (or dragarea) menu - the close event doesn't fire AT ALL

I swapped to using "beforeunload" instead of close - the window now closes but beforeunload isn't fired when closing via that menu either??

**Comment by bwin** [comment]
`win.close = function() {}` overwrites the close function of your `win`. You're using jquery-syntax. Attach listeners with `win.on` instead.

Don't know about different behaviour when closed from menu, though.

**Comment by shrewdlogarithm** [comment]
Ironically - I changed from "win.on(" because it's written as "win.close =
" in the docs ;0

**Comment by bwin** [comment]
In https://github.com/atom/atom-shell/blob/master/docs/api/browser-window.md (line 10) it's with `.on`. Couldn't find it as `win.close` through search. Maybe you we're looking at an old version and it is already fixed?

Are you still experiencing different behavior when closed from the menu?

**Comment by zcbenz** [comment]
You should use [close](https://github.com/atom/atom-shell/blob/master/docs/api/browser-window.md#event-close) event for this, it is not supported to override existing APIs in atom-shell.

**Comment by shrewdlogarithm** [comment]
First - the line in the docs which inspired my change of event assigned is this

    window.onbeforeunload = function(e) {

I've switched back to win.on("close",function()... syntax  and I'm still having issues - the main issue being that the event either never fires or at least never completes (and throws no errors and cannot be debugged in the devtools because the Window closes before it can do that)

I've tried the following code (intended to save window size/position) with both 'close' and 'beforeunload' events - it NEVER fires with the beforeunload, it SOMETIMES runs the first 'closing' message when run via 'close' - but nothing else (no errors - nothing else happens)

win is assigned from remote.getCurrentWindow() at window creation time.

````
win.on("close",function() {
            debugger // this won't work - window closes before debugger catches it???
            try {
                console.log("closing")
                if (win.isMaximized())
                    win.unMaximize()
                else if (win.isMinimized())
                    win.restore()
                console.log("closing2")
                console.log(win.getPosition())
                console.log(win.getSize())
                console.log("closing3")
                PREFS.setPref("COREwinx", win.getPosition()[0])
                PREFS.setPref("COREwiny", win.getPosition()[1])
                PREFS.setPref("COREwinwidth", win.getSize()[0])
                PREFS.setPref("COREwinheight", win.getSize()[1])
                console.log("done")
            } catch (e) {
                console.log("Error")
                console.log(e)
            }
        })
````

Something is not working as intended - note that beforeunload isn't ideal anyway and I don't really want to run this on reload...

p.s. this happens however I close the window - via win.close() or right-clicking the menu or even from main.js using app.quit()

Note also that app.quit() has it's own issues, including throwing a LOT of errors about missing callbacks and other problems - it's NOT the way to close an active app at this point...

**Comment by shrewdlogarithm** [comment]
OK - I've narrowed it down and I think I know what the problem is - 

````
    win.on("close",function() {
            try {
                console.log("closing") // this o
                if (win.isMaximized()) // this silently fails perhaps because you cannot reference 'win' at this point??
                  ... // so nothing else runs from here onwards
            } catch (e) {
                // but oddly, this doesn't run either???
                console.log("Error")
                console.log(e)
            }
        })
````

I suspect this relates to the errors you sometimes get if you reload a page/use app.quit() - you get errors to non-existant functions/callbacks which belong to 'win' (remote.getCurrentWindow() in this case)

What I've found is using

````
window.onbeforeunload = function() {
    // "win" is accessible in here just fine
    // just bear-in-mind this will run on reload as well as close
}

**Comment by shrewdlogarithm** [comment]
I've uploaded a test project to demo this (and another issue I'll raise separately re: Tray in OSX)

https://github.com/shrewdlogarithm/atom-shell-crash-demo

If that helps...

**Comment by bwin** [comment]
In the close-event it's definitely too late to do something window-related. Imagine a server: When you get a close-event on a connection you can't send anymore because the connection is closed.
I understand the confusion though (if you come from nw.js), as they handle their close-events differently (and even allow cancelling it from within the close-event).
The beforeunload-event is the way to go in your case.
The caveeat however, as you wrote (and I didn't knew) is that
> this will run on reload as well as close

**Comment by shrewdlogarithm** [comment]
The documentation implies that it's possible to prevent a window from
closing in the "close" event and it occurs BEFORE beforeunload fires too!

You can prevent the window closing in beforeunload - so you should be able
to in close (the docs imply this is the case)

So this is clearly a bug - win should not be unassigned or unavailable
during 'close' because you may cancel the event - and indeed want to refer
to it in beforeunload or even unload!??!?

tl;dr version

Referencing remote.getCurrentWindow() in win.on("close" fails (silently) -
but works in window.beforeunload DESPITE that running later!?

**Comment by bwin** [comment]
:sob: I'm sorry, I confused `close` and `closed`... I was talking about close**d**, but you were asking about close.
